Add basic Ledger hardware wallet support

lrettig opened the pull request:
Closes #44
Closes #10

lrettig committed:
Add basic ledger wallet support

Update upstream dependencies
Don't print master key by default
Allow empty private keys and mnemonics

lrettig committed:
Bugfixes and misc. improvements

Correctly print missing privkey, set key names correctly when using
ledger
Don't prompt twice on device when creating keys

lrettig committed:
Remove musl toolchain

Add required dylibs on Linux

lrettig commented:
Tests will fail until smkeys new module has been merged in https://github.com/spacemeshos/smkeys/pull/4

fasmat commented on wallet/bip32.go:
```diff
@@ -70,15 +79,49 @@ func NewMasterKeyPair(seed []byte) (*EDKeyPair, error) {
 func (kp *EDKeyPair) NewChildKeyPair(seed []byte, childIdx int) (*EDKeyPair, error) {
 	path := kp.Path.Extend(BIP44HardenedAccountIndex(uint32(childIdx)))
-	key, err := smbip32.Derive(HDPathToString(path), seed)
+	if kp.KeyType == typeLedger {
```
Imo a good place for a `switch` 🙂

fasmat commented on wallet/bip32.go:
```diff
@@ -70,15 +79,49 @@ func NewMasterKeyPair(seed []byte) (*EDKeyPair, error) {
+		return &EDKeyPair{
+			DisplayName: fmt.Sprintf("Child Key %d", childIdx),
+			Created:     common.NowTimeString(),
+			Private:     key[:],
```
Here we are converting the array back to a slice - I think we can return the key as a slice in `Derive` 🙂

fasmat commented on wallet/bip32.go:
```diff
@@ -7,6 +7,7 @@ import (
 	"fmt"
 
 	smbip32 "github.com/spacemeshos/smkeys/bip32"
+	ledger "github.com/spacemeshos/smkeys/remote-wallet"
```
you can add the other module by explicitly referencing the commit; are you using `go work` locally?

```
go get github.com/spacemeshos/smkeys@15e5e907
```

lrettig committed:
Factor in review comments

lrettig committed:
Attempt to fix CI workflow

Remove musl as it's no longer needed
Run golangci-lint through an action
Set necessary env vars in Makefile for staticcheck

lrettig committed:
Update dependencies

lrettig committed:
Ran go mod tidy

lrettig committed:
Fix linting

lrettig committed:
Fix up makefile recipes and GH workflow

fasmat commented on go.mod:
```diff
@@ -5,7 +5,7 @@ go 1.18
 	github.com/btcsuite/btcutil v1.0.2
 	github.com/jedib0t/go-pretty/v6 v6.4.6
-	github.com/spacemeshos/smkeys v1.0.2
+	github.com/spacemeshos/smkeys v1.0.3-0.20230511190607-5b2cfe71b4ba
```
Will there be a release of `smkeys` before the merge?

fasmat commented on Makefile:
```diff
@@ -121,42 +113,62 @@ tidy:
 
 .PHONY: test-tidy
-test-tidy:
+test-tidy: install
```
`install` was not a dependent target of most of the others because its not a `no-op` if it was already executed.

This will install the tools used multiple times in a CI run. using them locally also becomes more tedious.

lrettig replied:
fair point, will fix!

lrettig committed:
Update github.com/spacemeshos/smkeys dependency

lrettig committed:
Clean up Makefile recipes and CI workflow

Fix previous commit based on review feedback
